src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,1064 @@
+			if (rc != 0) {
+				D_GOTO(out, rc);
+			}
+		} else if (type == DAOS_PROP_CO_ACL) {
+			entry = &prop_query->dpp_entries[i];
+			rc = serialize_acl(file_id, entry, prop_str);
+			if (rc != 0) {
+				D_GOTO(out, rc);
+			}
+		} else if (type == DAOS_PROP_CO_LAYOUT_TYPE ||
+			   type == DAOS_PROP_CO_LAYOUT_VER ||
+			   type == DAOS_PROP_CO_CSUM ||
+			   type == DAOS_PROP_CO_CSUM_CHUNK_SIZE ||
+			   type == DAOS_PROP_CO_CSUM_SERVER_VERIFY ||
+			   type == DAOS_PROP_CO_REDUN_FAC ||
+			   type == DAOS_PROP_CO_REDUN_LVL ||
+			   type == DAOS_PROP_CO_SNAPSHOT_MAX ||
+			   type == DAOS_PROP_CO_COMPRESS ||
+			   type == DAOS_PROP_CO_ENCRYPT ||
+			   type == DAOS_PROP_CO_DEDUP ||
+			   type == DAOS_PROP_CO_DEDUP_THRESHOLD ||
+			   type == DAOS_PROP_CO_ALLOCED_OID) {
+			entry = &prop_query->dpp_entries[i];
+			rc = serialize_uint(file_id, entry->dpe_val,
+					    prop_str);
+			if (rc != 0) {
+				D_GOTO(out, rc);
+			}
+		} else {
+			rc = -DER_INVAL;
+			D_ERROR("Serialization of this container property "
+				"%s type is not supported "DF_RC"\n",
+				prop_str, DP_RC(rc));
+		}
+	}
+out:
+	return rc;
+}
+
+int
+daos_cont_serialize_attrs(hid_t file_id, hid_t *usr_attr_memtype,
+			  int num_attrs, char **names, char **buffers,
+			  size_t *sizes)
+{
+	int		rc = 0;
+	hid_t		status = 0;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	if (num_attrs == 0) {
+		D_GOTO(out_no_attrs, rc);
+	}
+
+	if (names == NULL || buffers == NULL || sizes == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+
+	/* Create the user attribute data space */
+	dims[0] = num_attrs;
+	dspace = H5Screate_simple(1, dims, NULL);
+	if (dspace < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to create dataspace\n");
+		D_GOTO(out, rc);
+	}
+
+	/* Create the user attribute dataset */
+	dset = H5Dcreate(file_id, SERIALIZE_ATTR_DSET, *usr_attr_memtype,
+			 dspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+	if (dset < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to create dataset\n");
+		D_GOTO(out, rc);
+	}
+
+	/* Allocate space for all attributes */
+	D_ALLOC(attr_data, num_attrs * sizeof(struct usr_attr));
+	if (attr_data == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+
+	/* Set the data for all attributes */
+	for (i = 0; i < num_attrs; i++) {
+		attr_data[i].attr_name = names[i];
+		attr_data[i].attr_val.p = buffers[i];
+		attr_data[i].attr_val.len = sizes[i];
+	}
+
+	status = H5Dwrite(dset, *usr_attr_memtype, H5S_ALL,
+			  H5S_ALL, H5P_DEFAULT, attr_data);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to write to dataset\n");
+		D_GOTO(out, rc);
+	}
+out:
+	D_FREE(attr_data);
+	if (dset > 0)
+		H5Dclose(dset);
+	if (*usr_attr_memtype > 0)
+		H5Tclose(*usr_attr_memtype);
+	if (dspace > 0)
+		H5Sclose(dspace);
+out_no_attrs:
+	return rc;
+}
+
+int
+daos_cont_serialize_md(char *filename, daos_prop_t *props, int num_attrs,
+		       char **names, char **buffers, size_t *sizes)
```

Comment:
Since this function is public, any pointers that are dereferenced in this function need NULL checks. E.g. filename